presto-main/src/main/java/com/facebook/presto/operator/scalar/MapTransformValueFunction.java

```diff
@@ -89,28 +109,99 @@ public ScalarFunctionImplementation specialize(BoundVariables boundVariables, in
-            try {
-                transformedValue = function.invoke(key, value);
-            }
-            catch (Throwable throwable) {
-                throw Throwables.propagate(throwable);
-            }
-
-            keyType.appendTo(block, position, resultBuilder);
-            writeNativeValue(transformedValueType, resultBuilder, transformedValue);
+        CallSiteBinder binder = new CallSiteBinder();
+        MapType transformedMapType = new MapType(keyType, transformedValueType);
+        Class<?> keyJavaType = Primitives.wrap(keyType.getJavaType());
+        Class<?> valueJavaType = Primitives.wrap(valueType.getJavaType());
+        Class<?> transformedValueJavaType = Primitives.wrap(transformedValueType.getJavaType());
+
+        ClassDefinition definition = new ClassDefinition(
+                a(PUBLIC, FINAL),
+                makeClassName("MapTransformValue"),
+                type(Object.class));
+        definition.declareDefaultConstructor(a(PRIVATE));
+
+        // define transform method
+        Parameter block = arg("block", Block.class);
+        Parameter function = arg("function", MethodHandle.class);
+        MethodDefinition method = definition.declareMethod(
+                a(PUBLIC, STATIC),
+                "transform",
+                type(Block.class),
+                ImmutableList.of(block, function));
+
+        BytecodeBlock body = method.getBody();
+        Scope scope = method.getScope();
+        Variable positionCount = scope.declareVariable(int.class, "positionCount");
+        Variable position = scope.declareVariable(int.class, "position");
+        Variable blockBuilder = scope.declareVariable(BlockBuilder.class, "blockBuilder");
+        Variable keyElement = scope.declareVariable(keyJavaType, "keyElement");
+        Variable valueElement = scope.declareVariable(valueJavaType, "valueElement");
+        Variable transformedValueElement = scope.declareVariable(transformedValueJavaType, "transformedValueElement");
+
+        // invoke block.getPositionCount()
+        body.append(positionCount.set(block.invoke("getPositionCount", int.class)));
+
+        // create the interleaved block builder
+        body.append(blockBuilder.set(newInstance(
+                InterleavedBlockBuilder.class,
+                constantType(binder, transformedMapType).invoke("getTypeParameters", List.class),
+                newInstance(BlockBuilderStatus.class),
+                positionCount)));
+
+        SqlTypeBytecodeExpression keySqlType = constantType(binder, keyType);
+        BytecodeNode loadKeyElement;
+        if (!keyType.equals(UNKNOWN)) {
+            // key element must be non-null
+            loadKeyElement = new BytecodeBlock().append(keyElement.set(keySqlType.getValue(block, position).cast(keyJavaType)));
         }
-        return resultBuilder.build();
+        else {
+            loadKeyElement = new BytecodeBlock().append(keyElement.set(constantNull(keyJavaType)));
```

Comment:
I understand that the original code didn't handle this. Now that you have an explicit branch for UNKNOWN, it would seem reasonable that you generate a throw here. See #7508